[USMON-1010] unify http2 plaintext and TLS implementations

guyarb opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

The PR unifies plaintext and TLS implementations for HTTP2 monitoring.
The PR uses `pktbuf` as an abstraction layer to achieve maximum shared code between the plaintext and TLS versions

The PR removes redundant and duplicated functions.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Remove irrelevant code base and achieve single logic between plain http2 and TLS http2.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

#### Complexity changes
```
| Filename/Program                         |   Stack Usage |   Instructions Processed |   Instructions Processed limit |   Max States per Instruction |   Peak States |   Total States |
|------------------------------------------|---------------|--------------------------|--------------------------------|------------------------------|---------------|----------------|
| usm/socket__http2_dynamic_table_cleaner  |            -8 |                       -9 |                              0 |                            0 |            -1 |             -1 |
| usm/socket__http2_eos_parser             |            -8 |                       -1 |                              0 |                            0 |             0 |              0 |
| usm/socket__http2_filter                 |            16 |                   143854 |                              0 |                          -11 |          1753 |           2459 |
| usm/socket__http2_handle_first_frame     |             8 |                        3 |                              0 |                            0 |             4 |              3 |
| usm/socket__http2_headers_parser         |           -40 |                  -115166 |                              0 |                           -3 |           183 |          -3779 |
| usm/uprobe__http2_dynamic_table_cleaner  |             0 |                      -17 |                              0 |                            0 |            -1 |             -1 |
| usm/uprobe__http2_tls_eos_parser         |             8 |                        1 |                              0 |                            0 |             0 |              0 |
| usm/uprobe__http2_tls_filter             |            24 |                   136782 |                              0 |                           17 |          2598 |           4524 |
| usm/uprobe__http2_tls_handle_first_frame |            24 |                     -105 |                              0 |                            0 |            -6 |             -7 |
| usm/uprobe__http2_tls_headers_parser     |            24 |                   124410 |                              0 |                            4 |         -4611 |           3820 |
| usm/uprobe__http2_tls_termination        |             0 |                        0 |                              0 |                            0 |             0 |              0 |
```

#### Perf report

##### TLS
No CPU change
![image](https://github.com/DataDog/datadog-agent/assets/17148247/99b56b6f-4867-4a64-8501-819d303788a6)

No memory change 
![image](https://github.com/DataDog/datadog-agent/assets/17148247/5559249f-6a83-48b1-bfb4-7bf2ffe5f854)

Accuracy improved
![image](https://github.com/DataDog/datadog-agent/assets/17148247/925952ff-f5e1-4857-a959-2777a5b45cc3)

##### Plaintext

CPU slight reduction

![image](https://github.com/DataDog/datadog-agent/assets/17148247/f4f46255-06ea-4739-bfb4-d5398bce9f74)

Memory slight reduction

![Uploading image.png…]()

Accuracy didn't change

![image](https://github.com/DataDog/datadog-agent/assets/17148247/2ea3091c-a583-40d1-bf13-b72d455cae0d)


<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

guyarb committed:
usm: http2: handle_first_frame: Remove difference between TLS and plaintext versions

Zeroing filter_iterations and data_off of the iteration value in the TLS version.
Simplifying the copying of the plaintext dispatcher arguments

guyarb committed:
usm: http2: handle_first_frame: Migrate skip_preface

The commit introduces a common implementation for skip_preface, and uses it in plain and tls versions.
The commit removes tls_skip_preface as it is no longer needed.
The commit does not remove skip_preface as it is still needed by the classification code.

guyarb committed:
usm: http2: handle_first_frame: Migrate access data_off and data_end to be agnostic to implementation

The commit replaces every location in which we directly access data_off and data_end in
handle_first_frame implementations

guyarb committed:
usm: http2: handle_first_frame: Unify getting telemetry

The commit introduces a utility function to get the correct telemetry object
from the right map (plaintext or tls)

guyarb committed:
usm: http2: handle_first_frame: Unify loading bytes

guyarb committed:
usm: http2: handle_first_frame: Convert get_first_frame and fix_header_frame to pktbuf

Both helper methods now support pktbuf notation.
The commit also removed the plaintext and TLS implementations of the helper functions

guyarb committed:
usm: http2: handle_first_frame: Unify tail_call operations

Both implementations now use a unified way to call for the next program

guyarb committed:
usm: http2: handle_first_frame: Unify implementations

Introducing a common function to perform the logic of handle_first_frame.

guyarb committed:
usm: http2: eos parser: migrate program

guyarb committed:
usm: http2: dynamic table cleaner: migrate program

guyarb committed:
usm: http2: filter: preliminary change to reduce difference between TLS and plaintext

Instead of copying a copy of the entire skb_info, we simply store the
original value of the data offset, and we do the same in the TLS version.

The commit meant to reduce the difference between the TLS and plaintext versions
before the migration to pktbuf

guyarb committed:
usm: http2: filter: Unify telemetry extraction

guyarb committed:
usm: http2: filter: Unify extraction of data_off and data_end

guyarb committed:
usm: http2: filter: Unify loading bytes

guyarb committed:
usm: http2: filter: Unify tail_call operations

guyarb committed:
usm: http2: filter: Unify updating http2_iterations map before calling the next program

guyarb committed:
usm: http2: filter: convert find_relevant_frames to support pktbuf

The commit converts find_relevant_frames to pktubf_find_relevant_frames and supports pktbuf.
We remove tls_find_relevant_frames as it is no longer needed.

Also, the TLS implementation now supports multi-calling for the fitler program

guyarb committed:
usm: http2: filter: Move program logic into a shared function

guyarb committed:
usm: http2: parser: preliminary change to reduce difference between TLS and plaintext

Instead of copying a copy of the entire skb_info, we simply store the
original value of the data offset, and we do the same in the TLS version.

The commit meant to reduce the difference between the TLS and plaintext versions
before the migration to pktbuf

guyarb committed:
usm: http2: parser: Unify telemetry extraction

guyarb committed:
usm: http2: parser: Unify access to http2_iterations map

guyarb committed:
usm: http2: parser: Unify extraction of data_off and data_end

guyarb committed:
usm: http2: parser: Unify tail_call operations

guyarb committed:
usm: http2: parser: Migrate all helper functions to use pktbuf notation

This is a heavy commit, which converts all helper function requiered in headers_parser to use
pktbuf notation.

guyarb committed:
usm: http2: parser: Unify plaintext and TLS implementations

guyarb committed:
usm: http2: tests: Added simple test to load http2 binary

Yumasi commented on pkg/network/ebpf/c/protocols/http2/decoding.h:
```diff
@@ -63,28 +171,52 @@ static __always_inline bool parse_field_literal(struct __sk_buff *skb, skb_info_
+    headers_to_process->new_dynamic_value_offset = pktbuf_data_offset(pkt);
     headers_to_process->new_dynamic_value_size = str_len;
     headers_to_process->is_huffman_encoded = is_huffman_encoded;
     // If the string len (`str_len`) is in the range of [0, HTTP2_MAX_PATH_LEN], and we don't exceed packet boundaries
     // (skb_info->data_off + str_len <= skb_info->data_end) and the index is kIndexPath, then we have a path header,
     // and we're increasing the counter. In any other case, we're not increasing the counter.
     *interesting_headers_counter += (str_len > 0 && str_len <= HTTP2_MAX_PATH_LEN);
```
This comment still references `skb_info`, it needs to be updated. 

Additionally, it seems to explain and list what the code until there does check, and restates what the previous comments in the function already explained. One should be able to understand all of that by just reading the code, so I am not sure what is the added value of this comment.

I suggest listing the conditions in function's comment, and maybe have the inner comments reference those conditions instead, if the code is not clear enough for it to be obvious.

guyarb replied:
moved to the function's documentation

Yumasi reviewed (changes requested):
Great to see the pktbuf in action! Not much to change appart from a few outdated comments, and nit on some variable names.

guyarb commented:
rebased on top of [guy.arbitman/USMON-1010-pktbuf-prep](https://github.com/DataDog/datadog-agent/pull/26242)
removed a commit that introduced a verifier issue

guyarb commented:
rebased on top of main

guyarb committed:
fixup: usm: http2: parser: Unify plaintext and TLS implementations

guyarb committed:
fixup: usm: http2: filter: Move program logic into a shared function

guyarb committed:
fixup: usm: http2: parser: Unify plaintext and TLS implementations

guyarb committed:
fixup: usm: http2: parser: Unify plaintext and TLS implementations

guyarb committed:
fixup: usm: http2: parser: Unify plaintext and TLS implementations

Simplify logic to avoid verifier error

guyarb committed:
usm: http2: tests: Remove TLS specific values

No longer needed to distinguish between plaintext and TLS results

guyarb committed:
fixup: usm: http2: filter: convert find_relevant_frames to support pktbuf

Fixing typo

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding.h:
```diff
@@ -577,26 +669,61 @@ int socket__http2_handle_first_frame(struct __sk_buff *skb) {
-        return 0;
+        return;
     }
     // Overriding the data_off field of the cached skb_info. The next prog will start from the offset of the next valid
     // frame.
```
done

guyarb committed:
usm: http2: Modified documentation

guyarb committed:
usm: http2: Rename arrays

guyarb commented:
/merge
